proxy_agent/src/shared_state.rs

```diff
@@ -108,13 +117,73 @@ impl Default for SharedState {
             telemetry_reader_shutdown: false,
             telemetry_logger_shutdown: false,
             telemetry_logger_status_message: UNKNOWN_STATUS_MESSAGE.to_string(),
+            mock_server_ip: None,   // test only
+            mock_server_port: None, // test only
             // service
             #[cfg(windows)]
             service_status_handle: None,
         }
     }
 }
 
+pub mod shared_state_wrapper {
+    use super::SharedState;
+    use std::sync::{Arc, Mutex};
+    use tokio_util::sync::CancellationToken;
+
+    pub fn get_runtime(
+        shared_state: Arc<Mutex<SharedState>>,
+    ) -> Option<Arc<Mutex<tokio::runtime::Runtime>>> {
+        shared_state.lock().unwrap().runtime.clone()
+    }
+
+    /// Shutdown the runtime and release the lock on runtime and shared_state
+    /// call it when need to stop the async tasks gracefully
+    pub fn shutdown_runtime(shared_state: Arc<Mutex<SharedState>>) {
+        let mut shared_state = shared_state.lock().unwrap();
+        let runtime = shared_state.runtime.clone();
+        shared_state.runtime = None;
+        drop(shared_state); // Release the lock on shared_state
+
+        if let Some(runtime) = runtime {
+            let runtime = runtime.lock().unwrap();
+            drop(runtime); // Release the lock on runtime
+        }
+    }
+
+    pub fn cancel_cancellation_token(shared_state: Arc<Mutex<SharedState>>) {
+        shared_state.lock().unwrap().cancellation_token.cancel();
+    }
+
+    pub fn get_cancellation_token(shared_state: Arc<Mutex<SharedState>>) -> CancellationToken {
+        shared_state.lock().unwrap().cancellation_token.clone()
+    }
+
+    /// Check the cancellation token and return error if it is cancelled
+    pub fn check_cancellation_token(
+        shared_state: Arc<Mutex<SharedState>>,
+        fn_name: &str,
+    ) -> std::io::Result<()> {
+        if shared_state
+            .lock()
+            .unwrap()
```

Comment:
Will this cause a panic if the lock cannot be unwrapped/acquired?

Reply:
this is by design; we keep all the static variables in this shared_state and use arc/mutex to guarantee the threads safe. if the unwrap failed, the application got into bad state, it has to be panic.